src/OpenTelemetry.Exporter.OpenTelemetryProtocol/PersistentStorage/DirectorySizeTracker.cs

```diff
@@ -0,0 +1,84 @@
+// <copyright file="DirectorySizeTracker.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+#nullable enable
+
+namespace OpenTelemetry.PersistentStorage.FileSystem;
+
+/// <summary>
+/// Tracks the available storage in a specified directory.
+/// </summary>
+internal sealed class DirectorySizeTracker
+{
+    private readonly long maxSizeInBytes;
+    private readonly string path;
+    private long directoryCurrentSizeInBytes;
+
+    public DirectorySizeTracker(long maxSizeInBytes, string path)
+    {
+        this.maxSizeInBytes = maxSizeInBytes;
+        this.path = path;
+        this.directoryCurrentSizeInBytes = CalculateFolderSize(path);
+    }
+
+    public void FileAdded(long fileSizeInBytes) => Interlocked.Add(ref this.directoryCurrentSizeInBytes, fileSizeInBytes);
+
+    public void FileRemoved(long fileSizeInBytes) => Interlocked.Add(ref this.directoryCurrentSizeInBytes, fileSizeInBytes * -1);
+
+    public bool IsSpaceAvailable(out long currentSizeInBytes)
+    {
+        currentSizeInBytes = Interlocked.Read(ref this.directoryCurrentSizeInBytes);
+        return currentSizeInBytes < this.maxSizeInBytes;
```

Comment:
Technically, this isn't thread-safe. As the value of `this.directoryCurrentSizeInBytes` could change before the evaluating the result of `return currentSizeInBytes < this.maxSizeInBytes;`.

Do we know if this could be an issue based on our knowledge of the usage of these methods?

Reply:
Good catch. some inaccuracy is acceptable in this case. False positive is ok - As the file write will eventually fail if there is no space left.
False Negative is also ok - Writing to disk can be retried based on the need. 

This was done in order to avoid acquiring lock while writing/deleting.

Reply:
Got it. It would be good to add this as code comments.

Reply:
Sure, I would like to make changes upstream first and copy it here later to keep the copies in sync.

Reply:
> Got it. It would be good to add this as code comments.

+1

Reply:
updated